### PR TITLE
Finalize client

### PR DESCRIPTION
**Pull request by maxsmythe**
This change finalizes the client interface and add end-to-end tests for both the local and remote Rego drivers.

### CODE REVIEW

File: constraint/pkg/client/client.go [hunk 1]
```diff
@@ -266,52 +357,60 @@ func (c *client) getConstraintEntry(constraint *unstructured.Unstructured, lock
...
+		resp.BasicError = err
+		return resp
 	}
 	for _, target := range entry.Targets {
 		path, err := createConstraintPath(target, constraint)
 		// If we ever create multi-target constraints we will need to handle this more cleverly.
 		// the short-circuiting question, cleanup, etc.
 		if err != nil {
-			return err
+			resp.Errors[target] = err
```
[the remark this answers is not in the document]
Comment (maxsmythe) [reply constraint/pkg/client/client.go]: Good question.

My thought was that if a constraint specifically references multiple targets, we may want it to be "transactional" (enforced for all relevant targets or none), and avoid split state.

My thought for data was we have less context on its purpose and therefore have no ability to assume what recovery looks like.

I haven't thought through it deeply enough to have a strong opinion. LMK if you do.

File: constraint/pkg/client/client.go [hunk 2]
```diff
@@ -19,37 +21,93 @@ import (
...
+	AddConstraint(context.Context, *unstructured.Unstructured) *TargetResponse
+	RemoveConstraint(context.Context, *unstructured.Unstructured) *TargetResponse
 	ValidateConstraint(context.Context, *unstructured.Unstructured) error
 
 	// Reset the state of OPA
 	Reset(context.Context) error
 
 	// Review makes sure the provided object satisfies all stored constraints
-	Review(context.Context, interface{}) ([]*types.Result, error)
+	Review(context.Context, interface{}) (types.Responses, *TargetResponse)
```
[the remark this answers is not in the document]
Comment (maxsmythe) [reply constraint/pkg/client/client.go]: Good point.

I'll have to think about this some, but I like the idea of merging the objects and then returning an error object for the second term.

People can then inspect the first return value if error is non-nil to get more context. It's slightly more idiomatic and addresses your error-oriented comment below.

WDYT?

File: constraint/pkg/client/client.go [hunk 3]
```diff
@@ -19,37 +21,93 @@ import (
...
+	}
+	return b.String()
+}
+
+type TargetResponse struct {
+	Handled map[string]bool
+	// Per-target errors
+	Errors ErrorMap
+	// Fundamental error in the request
+	BasicError error
```
[the remark this answers is not in the document]
Comment (maxsmythe) [reply constraint/pkg/client/client.go]: See above comment which I think may address this.

**Comment by maxsmythe** [comment]
Thanks for the review!

I'll have a follow-on CL to address your concerns soon. I discovered while writing the README that target names with dots in them are not currently handled. This seems important for "constraints.gatekeeper.sh" ;)

**Comment by maxsmythe** [comment]
Comments should be addressed by the latest push.

File: constraint/pkg/client/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsandall) [reply constraint/pkg/client/client.go]: I think the new API looks good and should simplify error handling in the more basic case (e.g., single target integration where errors are unrecoverable and just need to be logged/reported somehow). :+1:

**Pull request merged**